src/onnx/parse_einsum.cpp

```diff
@@ -0,0 +1,766 @@
+
+    instruction_ref unsqueeze_transpose(const onnx_parser::node_info& info,
+                                        int_mat& cur_pair,
+                                        instruction_ref op) const
+    {
+        int_vec unsq_axes;
+        std::vector<std::tuple<int, int>> perm;
+
+        for(auto i = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                unsq_axes.push_back(i);
+            else
+                perm.push_back({cur_pair[1][i], i});
+        }
+        op = info.add_instruction(make_op("unsqueeze", {{"axes", unsq_axes}}), op);
+
+        std::sort(perm.begin(), perm.end(), [](auto lhs, auto rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+
+        int_vec new_perm(cur_pair[1].size());
+        std::iota(new_perm.begin(), new_perm.end(), 0);
+
+        for(auto i = 0, p = 0; i < cur_pair[1].size(); ++i)
+        {
+            if(cur_pair[1][i] == -1)
+                continue;
+
+            new_perm[std::get<1>(perm[p++])] = i;
+        }
+
+        op = apply_transpose_op(info, op, new_perm, cur_pair[1]);
+
+        return op;
+    }
+
+    instruction_ref transpose_squeeze(const onnx_parser::node_info& info,
+                                      int_mat& cur_pair,
+                                      instruction_ref op,
+                                      int_vec row_output) const
+    {
+        std::vector<std::tuple<int, int>> perm;
+        int_vec sq;
+
+        for(auto i = 0; i < row_output.size(); ++i)
+        {
+            if(row_output[i] == -1)
+                sq.push_back(i);
+            else
+                perm.push_back({row_output[i], i});
+        }
+
+        std::sort(perm.begin(), perm.end(), [](auto lhs, auto rhs) {
+            return std::get<0>(lhs) < std::get<0>(rhs);
+        });
+
+        int_vec new_perm(cur_pair[1].size());
+        std::iota(new_perm.begin(), new_perm.end(), 0);
+
+        for(auto i = 0, p = 0; i < row_output.size(); ++i)
+        {
+            if(row_output[i] == -1)
+                continue;
+
+            new_perm[i] = std::get<1>(perm[p++]);
+        }
+
+        op = apply_transpose_op(info, op, new_perm, cur_pair[1]);
+
+        if(not sq.empty())
+        {
+            op = info.add_instruction(make_op("squeeze", {{"axes", sq}}), op);
+            // compute output row
+            for(int a : sq)
+                cur_pair[1][a] = -1;
+        }
+
+        return op;
+    }
+
+    bool is_transpose_identity(int_vec perm) const
+    {
+        for(auto i = 0u; i < perm.size(); ++i)
+            if(perm[i] != i)
+                return false;
+
+        return true;
+    }
+
+    instruction_ref apply_transpose_op(const onnx_parser::node_info& info,
+                                       instruction_ref op,
+                                       const int_vec& perm,
+                                       int_vec& row) const
+    {
+        if(is_transpose_identity(perm))
```

Comment:
I dont think you need to check for this. We already have a simplification to remove noop transposes later.

Reply:
Ok, that makes sense. I removed the check.